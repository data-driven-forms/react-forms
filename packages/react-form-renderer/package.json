{
  "name": "@data-driven-forms/react-form-renderer",
  "version": "2.8.0",
  "description": "React form renderer for data-driven-forms.",
  "main": "dist/cjs/index.js",
  "module": "dist/esm/index.js",
  "typings": "dist/cjs/index.d.ts",
  "files": [
    "dist/"
  ],
  "license": "Apache-2.0",
  "scripts": {
    "start": "webpack-dev-server --env dev --config ./config/webpack.config.js --open --hot",
    "build": "rollup -c ./rollup.config.js && yarn build:typings",
    "build:typings": "node ../../scripts/copy-files.js",
    "release": "semantic-release"
  },
  "repository": "git@github.com:data-driven-forms/react-forms.git",
  "keywords": [
    "react",
    "forms",
    "javascript"
  ],
  "devDependencies": {
    "@babel/core": "^7.2.2",
    "@babel/plugin-proposal-class-properties": "^7.1.0",
    "@babel/plugin-syntax-dynamic-import": "^7.0.0",
    "@babel/preset-env": "^7.1.6",
    "@babel/preset-react": "^7.0.0",
    "@semantic-release/git": "^7.0.5",
    "@semantic-release/npm": "^5.1.1",
    "babel-core": "^7.0.0-bridge.0",
    "babel-eslint": "9.x",
    "babel-jest": "^23.6.0",
    "babel-loader": "^8.0.4",
    "babel-plugin-lodash": "^3.3.4",
    "css-loader": "^1.0.1",
    "enzyme": "^3.7.0",
    "enzyme-adapter-react-16": "^1.7.0",
    "enzyme-to-json": "^3.3.4",
    "glob": "^7.1.6",
    "html-webpack-plugin": "^3.2.0",
    "mini-css-extract-plugin": "^0.4.4",
    "node-sass": "^4.10.0",
    "react": "^16.13.1",
    "react-dom": "^16.13.1",
    "regenerator-runtime": "^0.12.1",
    "resolve-url-loader": "^3.0.0",
    "rollup": "^1.23.1",
    "rollup-plugin-babel": "^4.3.3",
    "rollup-plugin-commonjs": "^10.1.0",
    "rollup-plugin-node-globals": "^1.4.0",
    "rollup-plugin-node-resolve": "^5.2.0",
    "rollup-plugin-replace": "^2.2.0",
    "rollup-plugin-sourcemaps": "^0.5.0",
    "rollup-plugin-terser": "^5.1.2",
    "sass-loader": "^7.1.0",
    "semantic-release": "^16.0.4",
    "style-loader": "^0.23.1",
    "url-loader": "^1.1.2",
    "webpack": "^4.25.1",
    "webpack-cli": "^3.1.2",
    "webpack-dev-server": "^3.1.10",
    "webpack-merge": "^4.1.4"
  },
  "dependencies": {
    "final-form": "^4.19.1",
    "final-form-arrays": "^3.0.2",
    "final-form-focus": "^1.1.2",
    "lodash": "^4.17.15",
    "prop-types": "^15.7.2",
    "react-final-form": "^6.4.0",
    "react-final-form-arrays": "^3.1.1"
  },
  "peerDependencies": {
    "react": "^16.13.1",
    "react-dom": "^16.13.1"
  },
  "postpublish": "export RELEASE_DEMO=true"
}
