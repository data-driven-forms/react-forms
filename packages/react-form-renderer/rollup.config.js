import nodeResolve from 'rollup-plugin-node-resolve';
import commonjs from 'rollup-plugin-commonjs';
import babel from 'rollup-plugin-babel';
import replace from 'rollup-plugin-replace';
import nodeGlobals from 'rollup-plugin-node-globals';
import { terser } from 'rollup-plugin-terser';
import typescript from '@rollup/plugin-typescript';
import glob from 'glob';
import path from 'path';
import sourcemaps from 'rollup-plugin-sourcemaps';

const outputPaths = [...glob.sync(path.resolve(__dirname, './src/files/!(*.d.ts)')), ...glob.sync(path.resolve(__dirname, './src/files/*.js'))];

const globals = {
  react: 'React',
  'react-dom': 'ReactDOM'
};

const babelOptions = {
  exclude: /node_modules/,
  runtimeHelpers: true,
  configFile: '../../babel.config.js'
};

const commonjsOptions = {
  ignoreGlobal: true,
  include: /node_modules/,
  namedExports: {
    './src/form-renderer/helpers': ['composeValidators']
  }
};

function onwarn(warning) {
  throw Error(warning.message);
}

const plugins = [
  nodeResolve(),
  babel(babelOptions),
  commonjs(commonjsOptions),
  nodeGlobals(),
  replace({ 'process.env.NODE_ENV': JSON.stringify('production') }),
  terser({
    keep_classnames: true,
    keep_fnames: true
  }),
  typescript({
    allowSyntheticDefaultImports: true
  }),
  sourcemaps()
];

export default [
  ...['cjs', 'esm'].map((env) => ({
    onwarn,
    input: ['./src/index.js', ...outputPaths],
    output: {
      dir: `./dist/${env}`,
      format: env,
      name: '@data-driven-forms/react-form-renderer',
      exports: 'named',
      globals,
      sourcemap: true
    },
    external: Object.keys(globals),
    plugins
  })),
  {
    onwarn,
    input: './src/index.js',
    output: {
      file: `./dist/umd/index.js`,
      format: 'umd',
      name: '@data-driven-forms/react-form-renderer',
      exports: 'named',
      globals,
      sourcemap: true
    },
    external: Object.keys(globals),
    plugins
  }
];
