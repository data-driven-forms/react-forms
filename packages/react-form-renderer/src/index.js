export { default as componentTypes } from './component-types';
export { default as composeValidators } from './compose-validators';
export { default as condition } from './condition';
export { default as dataTypes } from './data-types';
export { default as defaultSchemaValidator } from './default-schema-validator';
export { default as FieldArray } from './field-array';
export { default as FieldProvider } from './field-provider';
export { default as FormRenderer } from './form-renderer';
export { default as FormSpy } from './form-spy';
export { default as Form } from './form';
export { default as FormError } from './form-error';
export { default as parseCondition } from './parse-condition';
export { default as RendererContext } from './renderer-context';
export { default as schemaErrors } from './schema-errors';
export { default as useFieldApi } from './use-field-api';
export { default as useFormApi } from './use-form-api';
export { default as validatorMapper } from './validator-mapper';
export { default as validatorTypes } from './validator-types';
export { default as validators } from './validators';
export { default as wizardContext } from './wizard-context';
