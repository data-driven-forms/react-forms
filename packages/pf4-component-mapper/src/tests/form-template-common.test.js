import React from 'react';
import { FormRenderer, Form, FormError } from '@data-driven-forms/react-form-renderer';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';

import FormTemplate from '../form-template';
import RenderWithProvider from '../../../../__mocks__/with-provider';
import componentMapper from '../component-mapper';

describe('FormTemplate PF4 Common', () => {
  let initialProps;
  let ContextWrapper;
  let formOptions;

  beforeEach(() => {
    formOptions = {
      onSubmit: jest.fn(),
      onReset: jest.fn(),
      onCancel: jest.fn(),
      canReset: true,
      pristine: true,
      getState: jest.fn().mockImplementation(() => ({})),
    };
    ContextWrapper = ({ children, ...props }) => (
      <RenderWithProvider value={{ formOptions }}>
        <Form onSubmit={jest.fn()}>{() => children}</Form>
      </RenderWithProvider>
    );
    initialProps = {
      formFields: <div>Formfields</div>,
      schema: {},
    };
  });

  it('should render title', () => {
    initialProps = {
      ...initialProps,
      schema: { title: 'I am title' },
    };

    render(
      <ContextWrapper>
        <FormTemplate {...initialProps} />
      </ContextWrapper>
    );

    expect(screen.getByText('I am title')).toBeInTheDocument();
    expect(screen.getByText('Submit')).toBeInTheDocument();
    expect(screen.getByText('Cancel')).toBeInTheDocument();
  });

  it('should hide buttons', () => {
    render(
      <ContextWrapper>
        <FormTemplate {...initialProps} showFormControls={false} />
      </ContextWrapper>
    );

    expect(() => screen.getByText('Submit')).toThrow();
    expect(() => screen.getByText('Cancel')).toThrow();
  });

  it('should render description', () => {
    initialProps = {
      ...initialProps,
      schema: { description: 'I am description' },
    };

    render(
      <ContextWrapper>
        <FormTemplate {...initialProps} />
      </ContextWrapper>
    );

    expect(screen.getByText('I am description')).toBeInTheDocument();
  });

  it('should render custom formControls', () => {
    const FormButtons = () => <div>Form Controls</div>;

    render(
      <ContextWrapper>
        <FormTemplate {...initialProps} FormButtons={FormButtons} />
      </ContextWrapper>
    );

    expect(screen.getByText('Form Controls')).toBeInTheDocument();
    expect(() => screen.getByText('Submit')).toThrow();
    expect(() => screen.getByText('Cancel')).toThrow();
  });

  it('should render all controls and with default labels', () => {
    const { container } = render(
      <ContextWrapper>
        <FormTemplate {...initialProps} />
      </ContextWrapper>
    );
    expect([...container.getElementsByTagName('button')].map((b) => b.textContent)).toEqual(['Submit', 'Cancel']);
  });

  it('should render only submit button', () => {
    const { container } = render(
      <ContextWrapper>
        <FormTemplate {...initialProps} canReset={false} onCancel={undefined} />
      </ContextWrapper>
    );
    expect([...container.getElementsByTagName('button')].map((b) => b.textContent)).toEqual(['Submit']);
  });

  it('should render buttons in correct order', () => {
    const { container } = render(
      <ContextWrapper>
        <FormTemplate {...initialProps} canSubmit canReset buttonOrder={['cancel', 'submit', 'reset']} />
      </ContextWrapper>
    );

    expect([...container.getElementsByTagName('button')].map((b) => b.textContent)).toEqual(['Cancel', 'Submit', 'Reset']);
  });

  it('should add missing buttons if not defined in button order', () => {
    const { container } = render(
      <ContextWrapper>
        <FormTemplate {...initialProps} canSubmit canReset buttonOrder={[]} />
      </ContextWrapper>
    );

    expect([...container.getElementsByTagName('button')].map((b) => b.textContent)).toEqual(['Submit', 'Reset', 'Cancel']);
  });

  it('should call cancel with form values', async () => {
    const expectedValues = { name: 'some name', b: { type: 'good' } };

    const onCancel = jest.fn();

    render(
      <FormRenderer
        onCancel={(values) => onCancel(values)}
        schema={{ fields: [] }}
        onSubmit={jest.fn}
        FormTemplate={FormTemplate}
        initialValues={expectedValues}
        componentMapper={{}}
      />
    );

    await userEvent.click(screen.getByText('Cancel'));

    expect(onCancel).toHaveBeenCalledWith(expectedValues);
  });

  it('show form alert message', async () => {
    render(
      <FormRenderer
        schema={{
          fields: [
            {
              component: 'text-field',
              name: 'field',
              'aria-label': 'text',
            },
          ],
        }}
        validate={({ field }) => {
          if (field) {
            return { [FormError]: 'some error title' };
          }
        }}
        onSubmit={jest.fn()}
        FormTemplate={FormTemplate}
        componentMapper={componentMapper}
      />
    );

    expect(() => screen.getByText('some error title')).toThrow();

    await userEvent.type(screen.getByLabelText('text'), 'something');

    expect(screen.getByText('some error title')).toBeInTheDocument();
  });

  it('show form alert message as object', async () => {
    render(
      <FormRenderer
        schema={{
          fields: [
            {
              component: 'text-field',
              name: 'field',
              'aria-label': 'text',
            },
          ],
        }}
        validate={({ field }) => {
          if (field) {
            return { [FormError]: { title: 'some error title', description: 'some description' } };
          }
        }}
        onSubmit={jest.fn()}
        FormTemplate={FormTemplate}
        componentMapper={componentMapper}
      />
    );

    expect(() => screen.getByText('some error title')).toThrow();
    expect(() => screen.getByText('some description')).toThrow();

    await userEvent.type(screen.getByLabelText('text'), 'something');

    expect(screen.getByText('some error title')).toBeInTheDocument();
    expect(screen.getByText('some description')).toBeInTheDocument();
  });
});
