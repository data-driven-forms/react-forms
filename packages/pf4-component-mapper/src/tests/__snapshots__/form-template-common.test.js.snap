// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FormTemplate PF4 Common should add missing buttons if not defined in button order 1`] = `
<ContextWrapper>
  <RenderWithProvider
    value={
      Object {
        "formOptions": Object {
          "canReset": true,
          "onCancel": [MockFunction],
          "onReset": [MockFunction],
          "onSubmit": [MockFunction],
          "pristine": true,
        },
      }
    }
  >
    <ReactFinalForm
      onSubmit={[Function]}
    >
      <ReactFinalForm
        onSubmit={[MockFunction]}
      >
        <PF4FormTemplate
          buttonOrder={Array []}
          canReset={true}
          canSubmit={true}
          formFields={
            <div>
              Formfields
            </div>
          }
          schema={Object {}}
        >
          <FormTemplate
            Button={[Function]}
            ButtonGroup={[Function]}
            Description={[Function]}
            FormWrapper={[Function]}
            Title={[Function]}
            buttonOrder={Array []}
            canReset={true}
            canSubmit={true}
            disableSubmit={Array []}
            formFields={
              <div>
                Formfields
              </div>
            }
            schema={Object {}}
            showFormControls={true}
          >
            <Component>
              <form
                className="pf-c-form"
                noValidate={true}
              >
                <div>
                  Formfields
                </div>
                <FormSpy>
                  <FormControls
                    Button={[Function]}
                    ButtonGroup={[Function]}
                    FormSpy={[Function]}
                    buttonOrder={Array []}
                    canReset={true}
                    canSubmit={true}
                    cancelLabel="Cancel"
                    formFields={
                      <div>
                        Formfields
                      </div>
                    }
                    formSpyProps={
                      Object {
                        "active": undefined,
                        "dirty": false,
                        "dirtyFields": Object {},
                        "dirtyFieldsSinceLastSubmit": Object {},
                        "dirtySinceLastSubmit": false,
                        "error": undefined,
                        "errors": Object {},
                        "form": Object {
                          "batch": [Function],
                          "blur": [Function],
                          "change": [Function],
                          "destroyOnUnregister": false,
                          "focus": [Function],
                          "getFieldState": [Function],
                          "getRegisteredFields": [Function],
                          "getState": [Function],
                          "initialize": [Function],
                          "isValidationPaused": [Function],
                          "mutators": Object {},
                          "pauseValidation": [Function],
                          "registerField": [Function],
                          "reset": [Function],
                          "resetFieldState": [Function],
                          "resumeValidation": [Function],
                          "setConfig": [Function],
                          "submit": [Function],
                          "subscribe": [Function],
                        },
                        "hasSubmitErrors": false,
                        "hasValidationErrors": false,
                        "initialValues": undefined,
                        "invalid": false,
                        "modified": Object {},
                        "pristine": true,
                        "submitError": undefined,
                        "submitErrors": undefined,
                        "submitFailed": false,
                        "submitSucceeded": false,
                        "submitting": false,
                        "touched": Object {},
                        "valid": true,
                        "validating": false,
                        "values": Object {},
                        "visited": Object {},
                      }
                    }
                    onCancel={[MockFunction]}
                    onReset={[MockFunction]}
                    resetLabel="Reset"
                    schema={Object {}}
                    submitLabel="Submit"
                  >
                    <ButtonGroup>
                      <Component>
                        <div
                          className="pf-c-form__group pf-m-action"
                        >
                          <div
                            className="pf-c-form__actions"
                          >
                            <Button
                              buttonType="submit"
                              key="form-submit"
                              label="Submit"
                              type="submit"
                              variant="primary"
                            >
                              <Component
                                type="submit"
                                variant="primary"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-primary"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  tabIndex={null}
                                  type="submit"
                                >
                                  Submit
                                </button>
                              </Component>
                            </Button>
                            <Button
                              buttonType="reset"
                              disabled={true}
                              key="form-reset"
                              label="Reset"
                              onClick={[MockFunction]}
                              type="button"
                            >
                              <Component
                                isDisabled={true}
                                onClick={[MockFunction]}
                                type="button"
                                variant="secondary"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-secondary"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={true}
                                  onClick={[MockFunction]}
                                  tabIndex={null}
                                  type="button"
                                >
                                  Reset
                                </button>
                              </Component>
                            </Button>
                            <Button
                              buttonType="cancel"
                              key="form-cancel"
                              label="Cancel"
                              onClick={[MockFunction]}
                              type="button"
                            >
                              <Component
                                onClick={[MockFunction]}
                                type="button"
                                variant="link"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-link"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  onClick={[MockFunction]}
                                  tabIndex={null}
                                  type="button"
                                >
                                  Cancel
                                </button>
                              </Component>
                            </Button>
                          </div>
                        </div>
                      </Component>
                    </ButtonGroup>
                  </FormControls>
                </FormSpy>
              </form>
            </Component>
          </FormTemplate>
        </PF4FormTemplate>
      </ReactFinalForm>
    </ReactFinalForm>
  </RenderWithProvider>
</ContextWrapper>
`;

exports[`FormTemplate PF4 Common should render all controls and with default labels 1`] = `
<ContextWrapper>
  <RenderWithProvider
    value={
      Object {
        "formOptions": Object {
          "canReset": true,
          "onCancel": [MockFunction],
          "onReset": [MockFunction],
          "onSubmit": [MockFunction],
          "pristine": true,
        },
      }
    }
  >
    <ReactFinalForm
      onSubmit={[Function]}
    >
      <ReactFinalForm
        onSubmit={[MockFunction]}
      >
        <PF4FormTemplate
          formFields={
            <div>
              Formfields
            </div>
          }
          schema={Object {}}
        >
          <FormTemplate
            Button={[Function]}
            ButtonGroup={[Function]}
            Description={[Function]}
            FormWrapper={[Function]}
            Title={[Function]}
            disableSubmit={Array []}
            formFields={
              <div>
                Formfields
              </div>
            }
            schema={Object {}}
            showFormControls={true}
          >
            <Component>
              <form
                className="pf-c-form"
                noValidate={true}
              >
                <div>
                  Formfields
                </div>
                <FormSpy>
                  <FormControls
                    Button={[Function]}
                    ButtonGroup={[Function]}
                    FormSpy={[Function]}
                    buttonOrder={
                      Array [
                        "submit",
                        "reset",
                        "cancel",
                      ]
                    }
                    canReset={false}
                    canSubmit={false}
                    cancelLabel="Cancel"
                    formFields={
                      <div>
                        Formfields
                      </div>
                    }
                    formSpyProps={
                      Object {
                        "active": undefined,
                        "dirty": false,
                        "dirtyFields": Object {},
                        "dirtyFieldsSinceLastSubmit": Object {},
                        "dirtySinceLastSubmit": false,
                        "error": undefined,
                        "errors": Object {},
                        "form": Object {
                          "batch": [Function],
                          "blur": [Function],
                          "change": [Function],
                          "destroyOnUnregister": false,
                          "focus": [Function],
                          "getFieldState": [Function],
                          "getRegisteredFields": [Function],
                          "getState": [Function],
                          "initialize": [Function],
                          "isValidationPaused": [Function],
                          "mutators": Object {},
                          "pauseValidation": [Function],
                          "registerField": [Function],
                          "reset": [Function],
                          "resetFieldState": [Function],
                          "resumeValidation": [Function],
                          "setConfig": [Function],
                          "submit": [Function],
                          "subscribe": [Function],
                        },
                        "hasSubmitErrors": false,
                        "hasValidationErrors": false,
                        "initialValues": undefined,
                        "invalid": false,
                        "modified": Object {},
                        "pristine": true,
                        "submitError": undefined,
                        "submitErrors": undefined,
                        "submitFailed": false,
                        "submitSucceeded": false,
                        "submitting": false,
                        "touched": Object {},
                        "valid": true,
                        "validating": false,
                        "values": Object {},
                        "visited": Object {},
                      }
                    }
                    onCancel={[MockFunction]}
                    onReset={[MockFunction]}
                    resetLabel="Reset"
                    schema={Object {}}
                    submitLabel="Submit"
                  >
                    <ButtonGroup>
                      <Component>
                        <div
                          className="pf-c-form__group pf-m-action"
                        >
                          <div
                            className="pf-c-form__actions"
                          >
                            <Button
                              buttonType="submit"
                              key="form-submit"
                              label="Submit"
                              type="submit"
                              variant="primary"
                            >
                              <Component
                                type="submit"
                                variant="primary"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-primary"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  tabIndex={null}
                                  type="submit"
                                >
                                  Submit
                                </button>
                              </Component>
                            </Button>
                            <Button
                              buttonType="cancel"
                              key="form-cancel"
                              label="Cancel"
                              onClick={[MockFunction]}
                              type="button"
                            >
                              <Component
                                onClick={[MockFunction]}
                                type="button"
                                variant="link"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-link"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  onClick={[MockFunction]}
                                  tabIndex={null}
                                  type="button"
                                >
                                  Cancel
                                </button>
                              </Component>
                            </Button>
                          </div>
                        </div>
                      </Component>
                    </ButtonGroup>
                  </FormControls>
                </FormSpy>
              </form>
            </Component>
          </FormTemplate>
        </PF4FormTemplate>
      </ReactFinalForm>
    </ReactFinalForm>
  </RenderWithProvider>
</ContextWrapper>
`;

exports[`FormTemplate PF4 Common should render buttons in correct order 1`] = `
<ContextWrapper>
  <RenderWithProvider
    value={
      Object {
        "formOptions": Object {
          "canReset": true,
          "onCancel": [MockFunction],
          "onReset": [MockFunction],
          "onSubmit": [MockFunction],
          "pristine": true,
        },
      }
    }
  >
    <ReactFinalForm
      onSubmit={[Function]}
    >
      <ReactFinalForm
        onSubmit={[MockFunction]}
      >
        <PF4FormTemplate
          buttonOrder={
            Array [
              "cancel",
              "submit",
              "reset",
            ]
          }
          canReset={true}
          canSubmit={true}
          formFields={
            <div>
              Formfields
            </div>
          }
          schema={Object {}}
        >
          <FormTemplate
            Button={[Function]}
            ButtonGroup={[Function]}
            Description={[Function]}
            FormWrapper={[Function]}
            Title={[Function]}
            buttonOrder={
              Array [
                "cancel",
                "submit",
                "reset",
              ]
            }
            canReset={true}
            canSubmit={true}
            disableSubmit={Array []}
            formFields={
              <div>
                Formfields
              </div>
            }
            schema={Object {}}
            showFormControls={true}
          >
            <Component>
              <form
                className="pf-c-form"
                noValidate={true}
              >
                <div>
                  Formfields
                </div>
                <FormSpy>
                  <FormControls
                    Button={[Function]}
                    ButtonGroup={[Function]}
                    FormSpy={[Function]}
                    buttonOrder={
                      Array [
                        "cancel",
                        "submit",
                        "reset",
                      ]
                    }
                    canReset={true}
                    canSubmit={true}
                    cancelLabel="Cancel"
                    formFields={
                      <div>
                        Formfields
                      </div>
                    }
                    formSpyProps={
                      Object {
                        "active": undefined,
                        "dirty": false,
                        "dirtyFields": Object {},
                        "dirtyFieldsSinceLastSubmit": Object {},
                        "dirtySinceLastSubmit": false,
                        "error": undefined,
                        "errors": Object {},
                        "form": Object {
                          "batch": [Function],
                          "blur": [Function],
                          "change": [Function],
                          "destroyOnUnregister": false,
                          "focus": [Function],
                          "getFieldState": [Function],
                          "getRegisteredFields": [Function],
                          "getState": [Function],
                          "initialize": [Function],
                          "isValidationPaused": [Function],
                          "mutators": Object {},
                          "pauseValidation": [Function],
                          "registerField": [Function],
                          "reset": [Function],
                          "resetFieldState": [Function],
                          "resumeValidation": [Function],
                          "setConfig": [Function],
                          "submit": [Function],
                          "subscribe": [Function],
                        },
                        "hasSubmitErrors": false,
                        "hasValidationErrors": false,
                        "initialValues": undefined,
                        "invalid": false,
                        "modified": Object {},
                        "pristine": true,
                        "submitError": undefined,
                        "submitErrors": undefined,
                        "submitFailed": false,
                        "submitSucceeded": false,
                        "submitting": false,
                        "touched": Object {},
                        "valid": true,
                        "validating": false,
                        "values": Object {},
                        "visited": Object {},
                      }
                    }
                    onCancel={[MockFunction]}
                    onReset={[MockFunction]}
                    resetLabel="Reset"
                    schema={Object {}}
                    submitLabel="Submit"
                  >
                    <ButtonGroup>
                      <Component>
                        <div
                          className="pf-c-form__group pf-m-action"
                        >
                          <div
                            className="pf-c-form__actions"
                          >
                            <Button
                              buttonType="cancel"
                              key="form-cancel"
                              label="Cancel"
                              onClick={[MockFunction]}
                              type="button"
                            >
                              <Component
                                onClick={[MockFunction]}
                                type="button"
                                variant="link"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-link"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  onClick={[MockFunction]}
                                  tabIndex={null}
                                  type="button"
                                >
                                  Cancel
                                </button>
                              </Component>
                            </Button>
                            <Button
                              buttonType="submit"
                              key="form-submit"
                              label="Submit"
                              type="submit"
                              variant="primary"
                            >
                              <Component
                                type="submit"
                                variant="primary"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-primary"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  tabIndex={null}
                                  type="submit"
                                >
                                  Submit
                                </button>
                              </Component>
                            </Button>
                            <Button
                              buttonType="reset"
                              disabled={true}
                              key="form-reset"
                              label="Reset"
                              onClick={[MockFunction]}
                              type="button"
                            >
                              <Component
                                isDisabled={true}
                                onClick={[MockFunction]}
                                type="button"
                                variant="secondary"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-secondary"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={true}
                                  onClick={[MockFunction]}
                                  tabIndex={null}
                                  type="button"
                                >
                                  Reset
                                </button>
                              </Component>
                            </Button>
                          </div>
                        </div>
                      </Component>
                    </ButtonGroup>
                  </FormControls>
                </FormSpy>
              </form>
            </Component>
          </FormTemplate>
        </PF4FormTemplate>
      </ReactFinalForm>
    </ReactFinalForm>
  </RenderWithProvider>
</ContextWrapper>
`;

exports[`FormTemplate PF4 Common should render only submit button 1`] = `
<ContextWrapper>
  <RenderWithProvider
    value={
      Object {
        "formOptions": Object {
          "canReset": true,
          "onCancel": [MockFunction],
          "onReset": [MockFunction],
          "onSubmit": [MockFunction],
          "pristine": true,
        },
      }
    }
  >
    <ReactFinalForm
      onSubmit={[Function]}
    >
      <ReactFinalForm
        onSubmit={[MockFunction]}
      >
        <PF4FormTemplate
          canReset={false}
          formFields={
            <div>
              Formfields
            </div>
          }
          schema={Object {}}
        >
          <FormTemplate
            Button={[Function]}
            ButtonGroup={[Function]}
            Description={[Function]}
            FormWrapper={[Function]}
            Title={[Function]}
            canReset={false}
            disableSubmit={Array []}
            formFields={
              <div>
                Formfields
              </div>
            }
            schema={Object {}}
            showFormControls={true}
          >
            <Component>
              <form
                className="pf-c-form"
                noValidate={true}
              >
                <div>
                  Formfields
                </div>
                <FormSpy>
                  <FormControls
                    Button={[Function]}
                    ButtonGroup={[Function]}
                    FormSpy={[Function]}
                    buttonOrder={
                      Array [
                        "submit",
                        "reset",
                        "cancel",
                      ]
                    }
                    canReset={false}
                    canSubmit={false}
                    cancelLabel="Cancel"
                    formFields={
                      <div>
                        Formfields
                      </div>
                    }
                    formSpyProps={
                      Object {
                        "active": undefined,
                        "dirty": false,
                        "dirtyFields": Object {},
                        "dirtyFieldsSinceLastSubmit": Object {},
                        "dirtySinceLastSubmit": false,
                        "error": undefined,
                        "errors": Object {},
                        "form": Object {
                          "batch": [Function],
                          "blur": [Function],
                          "change": [Function],
                          "destroyOnUnregister": false,
                          "focus": [Function],
                          "getFieldState": [Function],
                          "getRegisteredFields": [Function],
                          "getState": [Function],
                          "initialize": [Function],
                          "isValidationPaused": [Function],
                          "mutators": Object {},
                          "pauseValidation": [Function],
                          "registerField": [Function],
                          "reset": [Function],
                          "resetFieldState": [Function],
                          "resumeValidation": [Function],
                          "setConfig": [Function],
                          "submit": [Function],
                          "subscribe": [Function],
                        },
                        "hasSubmitErrors": false,
                        "hasValidationErrors": false,
                        "initialValues": undefined,
                        "invalid": false,
                        "modified": Object {},
                        "pristine": true,
                        "submitError": undefined,
                        "submitErrors": undefined,
                        "submitFailed": false,
                        "submitSucceeded": false,
                        "submitting": false,
                        "touched": Object {},
                        "valid": true,
                        "validating": false,
                        "values": Object {},
                        "visited": Object {},
                      }
                    }
                    onReset={[MockFunction]}
                    resetLabel="Reset"
                    schema={Object {}}
                    submitLabel="Submit"
                  >
                    <ButtonGroup>
                      <Component>
                        <div
                          className="pf-c-form__group pf-m-action"
                        >
                          <div
                            className="pf-c-form__actions"
                          >
                            <Button
                              buttonType="submit"
                              key="form-submit"
                              label="Submit"
                              type="submit"
                              variant="primary"
                            >
                              <Component
                                type="submit"
                                variant="primary"
                              >
                                <button
                                  aria-disabled={null}
                                  aria-label={null}
                                  className="pf-c-button pf-m-primary"
                                  data-ouia-component-id={null}
                                  data-ouia-component-type="PF4/Button"
                                  data-ouia-safe={true}
                                  disabled={false}
                                  tabIndex={null}
                                  type="submit"
                                >
                                  Submit
                                </button>
                              </Component>
                            </Button>
                          </div>
                        </div>
                      </Component>
                    </ButtonGroup>
                  </FormControls>
                </FormSpy>
              </form>
            </Component>
          </FormTemplate>
        </PF4FormTemplate>
      </ReactFinalForm>
    </ReactFinalForm>
  </RenderWithProvider>
</ContextWrapper>
`;
